### ISSUE TITLE
weird spacing with reveal

### ISSUE THREAD

**Issue by reagle** [astefanutti/decktape#151]
The following command results in a PDF with too much spacing below the title in some slides. 

```decktape reveal -size 960x700 http://reagle.org/joseph/talks/2018/05-ica-lifehack.html test.pdf```

For example, look at [slide #5, "Genres"](http://reagle.org/joseph/talks/2018/05-ica-lifehack.html#/genres). It's a 3 item numbered list, but item 3 is pushed off the bottom because of the spacing below "Genres".

[test.pdf](https://github.com/astefanutti/decktape/files/2005025/test.pdf)

**Comment by astefanutti** [comment]
I've already noticed some layout issues with images. Would you mind:
* Comparing with the PNG export by using the `--screenshots` option and check whether it's only affecting PDF output
* Creating a minimal test case so that we can report the issue to the Puppeteer project

**Comment by astefanutti** [comment]
Thanks. Indeed output directory has to exist. We need to produce a minimal test case, ideally removing  Decktape from the equation, and submit an issue to the Puppeteer project.

**Comment by reagle** [comment]
1. Why not have it automatically create directory if it doesn't exist?
2. Here's a [mimimal-er example](http://reagle.org/joseph/talks/2018/min-eg-decktape.html).

```
decktape reveal http://reagle.org/joseph/talks/2018/min-eg-decktape.html test.pdf
Loading page http://reagle.org/joseph/talks/2018/min-eg-decktape.html ...
Loading page finished with status: 200
Reveal JS plugin activated
```

Doesn't look like it has anything to do with images. (I don't know anything about decktape or Puppateer, just download decktape today in hope of using it.)

**Comment by astefanutti** [comment]
> Why not have it automatically create directory if it doesn't exist?

That's an improvement I want to do, though not high priority.

> Doesn't look like it has anything to do with images

It's more related to the page layout. I faced some similar issues with absolute positioning. We need to find a minimal HTML page that reproduces the issue, like GoogleChrome/puppeteer#915. But you can try reporting the issue directly to Puppeteer and we'll progress /refine from there.

**Comment by astefanutti** [comment]
No worries, I'll try to isolate the issue ASAP.

**Comment by umangv** [comment]
I'm facing exactly the same issue. It seems to have something to do with using `<ul>`'s and `<ol>`'s. 

On one page I have code that looks like the following and the space appears between the `<p>` and the `<ul>`.

    <h2>blah</h2>
    <p>blah</p>
    <ul>blah</ul>

**Comment by greatfireball** [comment]
I observed the same issue. On some of my `reveal.js` slides are a header and an image or a header and a list, but in pdf export the images/lists are moved at or below the bottom of the page. The screenshot export seems to be correct.

**Comment by Witee** [comment]
Has anyone get the solution?

**Comment by mgurov** [comment]
Same issue here. I can see it even when comparing the pdf rendering for the [second page](https://revealjs.com/#/1) of the standard reveal.js demo. The arrow images at the ["vertical slides"](https://revealjs.com/#/2) section are also very obviously wrongly positioned.

**Comment by mgurov** [comment]
![image](https://user-images.githubusercontent.com/339544/50057123-46f0f800-0166-11e9-96f0-3bb017d856ae.png)

**Comment by obilodeau** [comment]
Is there a workaround? Because besides decktape there is no alternative. reveal.js' `?print-pdf` feature is still broken. Some issues there were closed but I have different problems now...

**Comment by hvwaldow** [comment]
A workaround seems to be increasing (very much) the size. For example using ``--size='2048x1536'`` instead of ``--size='1024x768'`` works for me.

**Comment by AndiH** [comment]
Thanks, @hvwaldow, that was an easy fix for me!

**Comment by jonathan-g** [comment]
The workaround on increasing the size does not work for me, but using version 1.0.0 (which uses phantomjs instead of headless chrome and puppeteer) works fine: <https://github.com/astefanutti/decktape/releases/tag/v1.0.0>, so I would recommend just installing version 1.0 of decktape for anyone who needs to render reveal.js decks.

**Comment by kzorba** [comment]
Hello, I also have the same issue. Using v1.0.0 to bypass the problem. Any news regarding fix in the latest version?

**Comment by astefanutti** [comment]
@kzorba a minimal use case has to be built so that an issue can be reported in Chromium / Skia (the component used by the PDF engine). In the meantime, I'd be inclined to set the default size for Reveal.js to a larger value, as it seems to work-around the issue for some people.

**Comment by Peque** [comment]
As of today it seems `?print-pdf` is working as expected.

@hvwaldow's [workaround](https://github.com/astefanutti/decktape/issues/151#issuecomment-456166075) works too.

@astefanutti Can you help with creating a minimal reproducible example? I can create a reproducible example, but minimal as in https://github.com/puppeteer/puppeteer/issues/915 is out of my reach with my limited JS/HTML/CSS knowledge. :sweat_smile: :innocent:

**Comment by n-grounds** [comment]
I'm interested in this being fixed as well.  For now the over-sizing workaround seems to be working for me.

**Comment by DanySK** [comment]
@angelocroatti it looks like that's the root of our issue...

**Comment by dxps** [comment]
@astefanutti If it helps, I guess I'm facing the same issue (with all the details and sample included), described in an reveal-md [issue](https://github.com/webpro/reveal-md/issues/382). Is it?

**Comment by pykong** [comment]
As a variation of the aforementioned workaround via increasing the size option, I have found that not providing a size option at all can also be worth a trial. I do however set a screen size in my reveal.js configuration, though. 
Maybe this is of help to someone trying to work around the issue...

**Comment by nuhrberg** [comment]
In the meantime, is there a fix or other solutions? No workaround works for me. 
Currently, I have screenshots taken of the presentation and create a PDF file with those screenshots. However, the text cannot be highlighted and the quality is limited to the image resolution.

**Comment by jonathan-g** [comment]
One option is to roll back to version 1.0, which is the last time decktape worked reliably for reveal.js slides: https://github.com/astefanutti/decktape/releases/tag/v1.0.0

This isn't @astefanutti's fault. It's a bug with Chromium that needs to be fixed by the Chromium developers, and there's nothing that decktape can do about it.

Decktape needed to move to Chromium/Puppeteer because since 2018, PhantomJS is no longer maintained (https://github.com/ariya/phantomjs/issues/15344), so the old version of decktape that relied of PhantomJS won't be stable.

Basically, it's a bad situation, but there's not much any of us can do.

**Comment by jonathan-g** [comment]
Have you tried opening your reveal.js presentation in Chrome, adding `?print-pdf` at the end of the URL, and then printing from Chrome (Ctrl+P)? That is often the most reliable way to generate a PDF, at least for now.

**Comment by nuhrberg** [comment]
Thanks for the feedback. I didn't mean to blame anyone. I am very happy that I can create a PDF at all with decktape. 
I tried it via the Dockerfile which is included in the release of version 1.0.0. Unfortunately, there are very different spacing issues there. 
I create the Reveal presentation with [asciidoctor](https://docs.asciidoctor.org/reveal.js-converter/latest/) and use a header for my presentation there. Unfortunately this header is only displayed on the first page via `?print-pdf`, because it has an absolute position in the HTML/CSS. 
Anyway, thanks for the tips. I'll do some more trial and error. If I find a solution that works for me, I'll share it here.

**Comment by kappapiana** [comment]
> A workaround seems to be increasing (very much) the size. For example using `--size='2048x1536'` instead of `--size='1024x768'` works for me.

This actually saved my day. I had a very mixed presentation with backgrounds, footer and stuff, all I had was ?print-pdf and whatever combination I tried was useless (fixed one thing, two were broken), then I tried this, but the things that were broken magically went fixed, but other again went broken (like images pushed too low or too much space between some elements). By using `--size` I was able to have a perfect pdf export.

Note: since most presentations are 16:9, to render them the right ratio I safely use `--size='2048x1200'`, which is not perfect ratio, but proven to work and reliable in my experience. 

I wish I had found this before!

**Comment by rolandpoulter** [comment]
I've been able to mitigate spacing issues in the pdf output by switching from `margin` based spacing to using `padding` or event wrapping the misaligned element in a `div` with `relative` spacing and using `absolute` positioning to force it into the right position.

**Comment by astefanutti** [comment]
This should finally be fixed with the upgrade to Chromium 111, which contains support for printing via the new layout engine, a.k.a. LayoutNG, and some other related bug fixes:

https://chromestatus.com/feature/5108105013100544
https://bugs.chromium.org/p/chromium/issues/detail?id=1121942

**Comment by sadikyalcin** [comment]
I've had a ton of spacing issues with `3.4.1`. Updating to `3.10` fixed all of the issues. Thanks!
